src/main/java/com/amazon/opendistroforelasticsearch/sql/plugin/RestSqlAction.java

```diff
@@ -136,7 +140,61 @@ protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient cli
         return responseParams;
     }
 
-    private boolean isClientError(Exception e) {
+    private static void logAndPublishMetrics(final Exception e) {
+
+        if (isClientError(e)) {
+            Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_CUS).increment();
+        } else {
+            Metrics.getInstance().getNumericalMetric(MetricName.FAILED_REQ_COUNT_SYS).increment();
+        }
+        LOG.error(String.format("[%s] Failed during query execution", LogUtils.getRequestId()), e);
+    }
+
+    private static QueryAction explainRequest(final NodeClient client, final SqlRequest sqlRequest)
+            throws SQLFeatureNotSupportedException, SqlParseException {
+
+            final QueryAction queryAction = new SearchDao(client).explain(sqlRequest.getSql());
+            queryAction.setSqlRequest(sqlRequest);
+            return queryAction;
+    }
+
+    private void executeSqlRequest(final RestRequest request, final QueryAction queryAction,
+                                   final Client client, final RestChannel channel) throws Exception {
+
+        if (isExplainRequest(request)) {
+            final String jsonExplanation = queryAction.explain().explain();
+            channel.sendResponse(new BytesRestResponse(OK, jsonExplanation));
+        } else {
+            Map<String, String> params = request.params();
+            RestExecutor restExecutor = ActionRequestRestExecutorFactory.createExecutor(params.get("format"),
+                    queryAction);
+            //doing this hack because elasticsearch throws exception for un-consumed props
+            Map<String, String> additionalParams = new HashMap<>();
+            for (String paramName : responseParams()) {
+                if (request.hasParam(paramName)) {
+                    additionalParams.put(paramName, request.param(paramName));
+                }
+            }
+            restExecutor.execute(client, additionalParams, queryAction, channel);
+        }
+    }
+
+    private static boolean isExplainRequest(final RestRequest request) {
+
+        return request.path().endsWith("/_explain");
+    }
+
+    private static boolean explainRequiresBlockingIOCall(final SqlRequest request) {
+
+        return containsSubquery(request);
+    }
+
+    private static boolean containsSubquery(final SqlRequest request) {
+
+        return request.getSql().toLowerCase(Locale.ROOT).contains("(select ");
```

Comment:
how to handle "( select"? Is there any preprocessing already handle this?

Reply:
I can check for regex match, instead of hard-coded string, e.g.:
```
return request.getSql().toLowerCase(Locale.ROOT).matches("\\(\\s*select ");
```